module/state_synchronization/getter.go

```diff
@@ -0,0 +1,313 @@
+
+type executionDataGetterImpl struct {
+	serializer       *serializer
+	blobService      network.BlobService
+	maxBlobSize      int
+	maxBlobTreeDepth int
+	metrics          module.ExecutionDataGetterMetrics
+	logger           zerolog.Logger
+}
+
+// retrieveBlobs retrieves the blobs for the given CIDs from the blobservice, and sends them to the given BlobSender
+// in the order specified by the CIDs.
+func (s *executionDataGetterImpl) retrieveBlobs(parent context.Context, bs *blobs.BlobSender, cids []cid.Cid, logger zerolog.Logger) error {
+	defer bs.Close()
+
+	ctx, cancel := context.WithCancel(parent)
+	defer cancel()
+
+	blobChan := s.blobService.GetBlobs(ctx, cids) // initiate a batch request for the given CIDs
+	cachedBlobs := make(map[cid.Cid]blobs.Blob)
+	cidCounts := make(map[cid.Cid]int) // used to account for duplicate CIDs
+
+	for _, c := range cids {
+		cidCounts[c] += 1
+	}
+
+	for _, c := range cids {
+		blob, ok := cachedBlobs[c]
+
+		if !ok {
+			var err error
+
+			if blob, err = s.findBlob(ctx, blobChan, c, cachedBlobs, logger); err != nil {
+				return err
+			}
+		}
+
+		cidCounts[c] -= 1
+
+		if cidCounts[c] == 0 {
+			delete(cachedBlobs, c)
+		}
+
+		if err := bs.Send(blob); err != nil {
+			return fmt.Errorf("failed to send blob %v to blob channel: %w", blob.Cid(), err)
+		}
+	}
+
+	return nil
+}
+
+// findBlob retrieves blobs from the given channel, caching them along the way, until it either
+// finds the target blob or exhausts the channel.
+func (s *executionDataGetterImpl) findBlob(
+	ctx context.Context,
+	blobChan <-chan blobs.Blob,
+	target cid.Cid,
+	cache map[cid.Cid]blobs.Blob,
+	logger zerolog.Logger,
+) (blobs.Blob, error) {
+	targetLogger := logger.With().Str("target_cid", target.String()).Logger()
+	targetLogger.Debug().Msg("finding blob")
+
+	// Note: blobs are returned as they are found, in no particular order
+	for blob := range blobChan {
+		targetLogger.Debug().Str("cid", blob.Cid().String()).Msg("received blob")
+
+		// check blob size
+		blobSize := len(blob.RawData())
+
+		if blobSize > s.maxBlobSize {
+			return nil, &BlobSizeLimitExceededError{blob.Cid()}
+		}
+
+		cache[blob.Cid()] = blob
+
+		if blob.Cid() == target {
+			return blob, nil
+		}
+	}
+
+	// the blob channel may be closed as a result of the context being canceled,
+	// in which case we should return the context error.
+	if ctx.Err() != nil {
+		return nil, ctx.Err()
+	}
+
+	targetLogger.Debug().Msg("blob not found")
+
+	return nil, &BlobNotFoundError{target}
+}
+
+// getBlobs gets the given CIDs from the blobservice, reassembles the blobs, and deserializes the reassembled data into an object.
+//
+// blobs are fetched from the blobservice and sent over a blob channel as they arrive, to a separate goroutine, which performs the
+// deserialization in a streaming fashion.
+func (s *executionDataGetterImpl) getBlobs(ctx context.Context, cids []cid.Cid, logger zerolog.Logger) (interface{}, uint64, error) {
```

Comment:
Since the result is either a slice of CIDs or a ChunkExecutionData, better to make it clear here. if `isMoreCIDs` is true, read moreCIDs, if false, read `chunkExecutionData`.
```suggestion
func (s *executionDataGetterImpl) getBlobs(ctx context.Context, cids []cid.Cid, logger zerolog.Logger) (moreCIDs []cid.CID , chunkExecutionData *ChunkExecutionData, isChunkExecutionData bool, depth uint64, err error) {
```